consoleme/lib/aws.py

```diff
@@ -1262,3 +1274,204 @@ async def validate_iam_policy(policy: str, log_data: Dict):
+    all_org_structure = {}
+    for organization in config.get("cache_accounts_from_aws_organizations", []):
+        org_account_id = organization.get("organizations_master_account_id")
+        role_to_assume = organization.get(
+            "organizations_master_role_to_assume", config.get("policies.role_name")
+        )
+        if not org_account_id:
+            raise MissingConfigurationValue(
+                "Your AWS Organizations Master Account ID is not specified in configuration. "
+                "Unable to sync accounts from "
+                "AWS Organizations"
+            )
+
+        if not role_to_assume:
+            raise MissingConfigurationValue(
+                "ConsoleMe doesn't know what role to assume to retrieve account information "
+                "from AWS Organizations. please set the appropriate configuration value."
+            )
+        org_structure = await retrieve_org_structure(
+            org_account_id, region=config.region
+        )
+        all_org_structure.update(org_structure)
+    redis_key = config.get(
+        "cache_organization_structure.redis.key.org_structure_key", "AWS_ORG_STRUCTURE"
+    )
+    s3_bucket = None
+    s3_key = None
+    if config.region == config.get("celery.active_region", config.region) or config.get(
+        "environment"
+    ) in ["dev", "test"]:
+        s3_bucket = config.get("cache_organization_structure.s3.bucket")
+        s3_key = config.get("cache_organization_structure.s3.file")
+    await store_json_results_in_redis_and_s3(
+        all_org_structure,
+        redis_key=redis_key,
+        s3_bucket=s3_bucket,
+        s3_key=s3_key,
+    )
+    return all_org_structure
+
+
+async def _is_member_of_ou(
+    identifier: str, ou: Dict[str, Any]
+) -> Tuple[bool, Set[str]]:
+    """Recursively walk org structure to determine if the account or OU is in the org and, if so, return all OUs of which the account or OU is a member
+
+    Args:
+        identifier: AWS account or OU ID
+        ou: dictionary representing the organization/organizational unit structure to search
+    """
+    found = False
+    ou_path = set()
+    for child in ou.get("Children", []):
+        if child.get("Id") == identifier:
+            found = True
+        elif child.get("Type") == "ORGANIZATIONAL_UNIT":
+            found, ou_path = await _is_member_of_ou(identifier, child)
+        if found:
+            ou_path.add(ou.get("Id"))
+            break
+    return found, ou_path
+
+
+async def get_organizational_units_for_account(identifier: str) -> Set[str]:
+    """Return a set of Organizational Unit IDs for a given account or OU ID
+
+    Args:
+        identifier: AWS account or OU ID
+    """
+    all_orgs = await get_org_structure()
+    for org_id, org_structure in all_orgs.items():
+        found, organizational_units = await _is_member_of_ou(identifier, org_structure)
+        if found:
+            break
+    if not organizational_units:
```

Comment:
Should this be tabbed? 

organizational_units might be undef at this point

Reply:
Ah! I didn't want to log a warning on every iteration, but nice catch on the uninitialized var

Reply:
Hmm, maybe just set organizational_units to None on the top?

Reply:
Updated to initialize to an empty set, which is what `_is_member_of_ou()` returns by default (and thus this function). It's still falsy, so the logic is the name as initializing to `None` but the return type stays correct even if we have zero iterations of the `for` loop.